### ISSUE TITLE
Nested "bundle exec" setups invalid BUNDLE_BIN_PATH

### ISSUE THREAD

**Issue by jperville** [bundler/bundler#5621]
# Issue description

We have some legacy code which invoke `bundle exec somecommand` from within a script which itself is run by `bundle exec`; in other words, the `somecommand` command is actually invoked like this `bundle exec bundle exec somecommand`.

This behavior was working fine with bundler v1.12.6, but when we upgraded to bundler v1.14.6, it broke.  We got tools such as "compass" started complaining about not being to load "json_gem", despite the "json" gem being listed in the top Gemfile.

# Minimal test

Here is a small test to know whether a given BUNDLERVERSION is affected:
```sh
$ sudo gem install bundler -v $BUNDLERVERSION
/bin/bash -c 'bundle exec bundle exec env | grep BUNDLE_BIN_PATH | cut -d= -f2 | xargs ls -l'
```

# Automated test

I automated the test using Docker, in the following gist: https://gist.github.com/jperville/2a11806cfaa77269063a6b2121420fe6 . Checkout the above gist, then run `docker build -t test-bundler .` in the directory containing the `Dockerfile`.

Then run the following tests:

```sh
# bundler v1.12.6 => WORKS
$ docker run --rm -ti test-bundler 1.12.6
Fetching: bundler-1.12.6.gem (100%)
Successfully installed bundler-1.12.6
Parsing documentation for bundler-1.12.6
Installing ri documentation for bundler-1.12.6
Done installing documentation for bundler after 3 seconds
1 gem installed
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all
non-root users on this machine.
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Bundled gems are installed into ./.bundle.
testing bundler version 1.12.6 ...-rwxr-xr-x 1 root root 878 Apr 26 20:00 /var/lib/gems/2.3.0/gems/bundler-1.12.6/exe/bundle

# bundler v1.14.6 => BROKEN
$ docker run --rm -ti test-bundler 1.14.6
Fetching: bundler-1.14.6.gem (100%)
Successfully installed bundler-1.14.6
Parsing documentation for bundler-1.14.6
Installing ri documentation for bundler-1.14.6
Done installing documentation for bundler after 3 seconds
1 gem installed
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all
non-root users on this machine.
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Bundled gems are installed into ./.bundle.
testing bundler version 1.14.6 ...ls: cannot access '/var/lib/gems/2.3.0/gems/bundler-1.14.6/lib/gems/bundler-1.14.6/bin/bundle': No such file or directory
```

Using the above test, we find out that the regression appeared between v1.13.1 and v1.13.2 of bundler.

```sh
# bundler v1.13.1 => WORKS
$ docker run --rm -ti test-bundler 1.13.1
Fetching: bundler-1.13.1.gem (100%)
Successfully installed bundler-1.13.1
Parsing documentation for bundler-1.13.1
Installing ri documentation for bundler-1.13.1
Done installing documentation for bundler after 3 seconds
1 gem installed
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all
non-root users on this machine.
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Bundled gems are installed into ./.bundle.
testing bundler version 1.13.1 ...-rwxr-xr-x 1 root root 1161 Apr 26 20:01 /var/lib/gems/2.3.0/gems/bundler-1.13.1/exe/bundle

# bundler v1.13.2 => BROKEN
$ docker run --rm -ti test-bundler 1.13.2
Fetching: bundler-1.13.2.gem (100%)
Successfully installed bundler-1.13.2
Parsing documentation for bundler-1.13.2
Installing ri documentation for bundler-1.13.2
Done installing documentation for bundler after 3 seconds
1 gem installed
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all
non-root users on this machine.
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Bundled gems are installed into ./.bundle.
testing bundler version 1.13.2 ...ls: cannot access '/var/lib/gems/2.3.0/gems/bundler-1.13.2/lib/gems/bundler-1.13.2/bin/bundle': No such file or directory
```

**Comment by segiddins** [comment]
I'm not seeing this failing on either 1.14.6 nor 1.15.0.pre.2

```shell
$ GEM_REQUIREMENT_BUNDLER="1.14.6" /bin/bash -c 'bundle exec bundle exec env | grep BUNDLE_BIN_PATH | cut -d= -f2 | xargs ls -l'
-rwxr-xr-x  1 segiddins  staff  1113 Apr 15 18:48 /Users/segiddins/.gem/ruby/2.4.1/gems/bundler-1.14.6/exe/bundle
```

**Comment by jperville** [comment]
@segiddins can you test with the Dockerfile in the above gist? I have the issue on Linux platform (ubuntu and centos7) but never tried on Apple platform.

```sh
$ /bin/bash -c 'bundle exec bundle exec env | grep BUNDLE_BIN_PATH | cut -d= -f2 | xargs ls -l' 
ls: cannot access '/var/lib/gems/2.2.0/gems/bundler-1.14.5/lib/gems/bundler-1.14.5/bin/bundle': No such file or directory
```

**Comment by segiddins** [comment]
I don't have docker installed

**Comment by colby-swandale** [comment]
I can confirm the `BUNDLE_BIN_PATH` not being correct when running `bundle exec bundle exec env` but I can't replicate the gem load problem that was reported.

**Comment by jperville** [comment]
I will reproduce with a simplified version taken from the build  of legacy frontend app in the coming days.

**Comment by jperville** [comment]
@colby-swandale have a look at this gist : https://gist.github.com/jperville/857ff9ebaa81cdf3be3b5c1e6f601859

The gist shows how the "compass compile" command will work when run through "bundle exec" but fail because of loading issues when invoked with nested "bundle exec". The example is taken from the build process of one of our legacy webapps, and I trimmed everything not necessary to reproduce the issue.

```sh
git clone https://gist.github.com/857ff9ebaa81cdf3be3b5c1e6f601859.git nested-bundle-exec-compass
cd nested-bundle-exec-compass

docker build -t rubytestimage .

# first, compile the compass application with a single "bundle exec", it works as expected (despite the deprecation warnings) and produces the stylesheets/main.css file.
docker run --rm -ti rubytestimage /bin/bash -c "cd stylesheets && scl enable rh-ruby23 -- bundle exec compass compile --time"

# then, reproduce the bug with nested  "bundle exec compass"; it will fail because compass' multi_json dependency fails to automagically load the "json" gem despite that gem being listed in the bundle.
docker run --rm -ti rubytestimage /bin/bash -c "cd stylesheets && scl enable rh-ruby23 -- bundle exec bundle exec compass compile --time"

# to investigate, redo the above commands manually in a shell, then inspect the content of "stylesheets/main.css", it contains a stacktrace
```

**Comment by jules2689** [comment]
```shell
~/ ➜ ruby -e "puts Gem.activate_bin_path('bundler', 'bundle', '>= 0.a')"
/Users/juliannadeau/.gem/ruby/2.3.3/gems/bundler-1.15.0.pre.2/exe/bundle
```

```shell
~/ ➜ bundle exec ruby -e "puts Gem.activate_bin_path('bundler', 'bundle', '>= 0.a')"
/opt/rubies/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.15.0.pre.3/exe/bundle
```

```shell
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.11
  - RUBY VERSION: 2.3.3 (2016-11-21 patchlevel 222) [x86_64-darwin16]
  - INSTALLATION DIRECTORY: /Users/juliannadeau/.gem/ruby/2.3.3
  - USER INSTALLATION DIRECTORY: /Users/juliannadeau/.gem/ruby/2.3.0
  - RUBY EXECUTABLE: /opt/rubies/2.3.3/bin/ruby
  - EXECUTABLE DIRECTORY: /Users/juliannadeau/.gem/ruby/2.3.3/bin
  - SPEC CACHE DIRECTORY: /Users/juliannadeau/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /opt/rubies/2.3.3/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-darwin-16
  - GEM PATHS:
     - /Users/juliannadeau/.gem/ruby/2.3.3
     - /opt/rubies/2.3.3/lib/ruby/gems/2.3.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/juliannadeau/.gem/ruby/2.3.3/bin
     - /opt/rubies/2.3.3/lib/ruby/gems/2.3.0/bin
     - /opt/rubies/2.3.3/bin
     - /Users/juliannadeau/.nvm/versions/node/v5.7.0/bin
     - /Users/juliannadeau/.pyenv/shims
     - /usr/local/sbin
     - ~/bin
     - /usr/local/bin
     - /usr/local/git/bin
     - /user/local
     - /Applications/VMware Fusion.app/contents/Library
     - /Applications/Postgres.app/Contents/MacOS/bin
     - /Users/juliannadeau/.gem/ruby/2.3.1/bin/rubocop
     - /Users/juliannadeau/.pyenv/shims
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /usr/local/MacGPG2/bin
     - /Library/TeX/texbin
     - /Users/juliannadeau/golang/bin
```

**Comment by jules2689** [comment]
This seems to be caused by `BUNDLE_BIN_PATH` being set

**Comment by colby-swandale** [comment]
This issue is also the same issue as #6629 which was fixed by #6689.
